### ISSUE TITLE
Prettier misplaces commented out code between if else

### ISSUE THREAD

**Issue by dumptyd** [prettier/prettier#2573]
```js
let a = 2;

if (a==2) {
  console.log('2');
}
// else if (a==3) {
//   console.log('3');
// }
else if (a==4) {
  console.log('4');
}
```

becomes

```js
let a = 2;

if (a == 2) {
  console.log('2');
} else if (a == 4) {
  // else if (a==3) {
  // console.log('3');
  // }
  console.log('4');
}
```

[Live example](https://prettier.io/playground/#%7B%22content%22%3A%22let%20a%20%3D%202%3B%5Cn%5Cnif%20(a%3D%3D2)%20%7B%5Cn%20console.log(%272%27)%3B%5Cn%7D%5Cn%2F%2F%20else%20if%20(a%3D%3D3)%20%7B%5Cn%2F%2F%20console.log(%273%27)%3B%5Cn%2F%2F%20%7D%5Cnelse%20if%20(a%3D%3D4)%20%7B%5Cn%20console.log(%274%27)%3B%5Cn%7D%5Cn%22%2C%22options%22%3A%7B%22printWidth%22%3A80%2C%22tabWidth%22%3A2%2C%22singleQuote%22%3Atrue%2C%22trailingComma%22%3A%22none%22%2C%22bracketSpacing%22%3Atrue%2C%22jsxBracketSameLine%22%3Afalse%2C%22parser%22%3A%22babylon%22%2C%22semi%22%3Atrue%2C%22useTabs%22%3Afalse%2C%22doc%22%3Afalse%2C%22ast%22%3Afalse%7D%7D)

**Comment by azz** [comment]
This is a tricky one as prettier will always put ` else if {` on the same line as `}`.

**Comment by aboyton** [comment]
That said, if the code was already formatted with prettier before commenting out then it would presumably look like

```js
let a = 2;

if (a==2) {
 console.log('2');
} else if (a==3) {
  console.log('3');
} else if (a==4) {
 console.log('4');
}
```
which would be commented out like this
```js
let a = 2;

if (a==2) {
 console.log('2');
// } else if (a==3) {
//   console.log('3');
} else if (a==4) {
 console.log('4');
}
```
which is then stable. Thus the problem seems only to occur if you're running prettier for the first time on code that's got bits that are commented out, which while I'm sure happens is hopefully less frequently. Ideally [commented code doesn't exist in a repo](https://medium.com/@kentcdodds/please-don-t-commit-commented-out-code-53d0b5b26d5f) :)

**Comment by azz** [comment]
From #3084:

**Prettier 1.7.4**
[Playground link](https://prettier.io/playground/#%7B%22content%22%3A%22%2F%2F%20First%20condition%20comment.%5Cnif%20(firstCondition())%20%7B%5Cn%20%20doSomething()%3B%5Cn%7D%5Cn%2F%2F%20Second%20condition%20comment.%5Cnelse%20if%20(secondCondition())%20%7B%5Cn%20%20doSomethingElse()%3B%5Cn%7D%22%2C%22options%22%3A%7B%22ast%22%3Afalse%2C%22bracketSpacing%22%3Atrue%2C%22doc%22%3Afalse%2C%22jsxBracketSameLine%22%3Afalse%2C%22output2%22%3Afalse%2C%22parser%22%3A%22babylon%22%2C%22printWidth%22%3A80%2C%22semi%22%3Atrue%2C%22singleQuote%22%3Afalse%2C%22tabWidth%22%3A2%2C%22trailingComma%22%3A%22none%22%2C%22useTabs%22%3Afalse%7D%7D)

**Input:**
```jsx
// First condition comment.
if (firstCondition()) {
  doSomething();
}
// Second condition comment.
else if (secondCondition()) {
  doSomethingElse();
}
```

**Output:**
```jsx
// First condition comment.
if (firstCondition()) {
  doSomething();
} else if (secondCondition()) {
  // Second condition comment.
  doSomethingElse();
}

```

**Comment by azz** [comment]
Related: #807

**Comment by jagi** [comment]
I'm also putting here my modified comment from the #3084 issue explaining why it's wrong:

> Comments are always about describing current or the next line of code or block of code. Of course, as long as a comment's location is in the correct place. And I think that putting a comment as in my "Input" example is the best location for it. It's definitely much better than other possibilities.

**Solution 1:**
```js
if (firstCondition()) {
  doSomething();
} else if (secondCondition()) { // Comment can excide line width, then what?
  doSomethingElse();
}
```

**Solution 2:**
```js
if (firstCondition()) {
  doSomething();
  // Commenting next line but put int the brackets for the previous if statement?
  // And also not indented the same as the next line I'm commenting.
} else if (secondCondition()) {
  doSomethingElse();
}
```

**Solution 3:**
```js
if (firstCondition()) {
  doSomething();
} /* Does anyone do it like that? */ else if / (secondCondition()) {
  doSomethingElse();
}
```

**Solution 4:**
```js
if (firstCondition()) {
  doSomething();
} else if (secondCondition()) {
  // I'm already after line that I'm commenting. Am I commenting "doSomethingElse();"?
  doSomethingElse();
}
```

**IMHO Best Solution:**
```js
if (firstCondition()) {
  doSomething();
}
// Commenting next line and looking very clean
// and can be wrapped easily if too long.
else if (secondCondition()) {
  doSomethingElse();
}
```

Of course, it has its downsides like for example after closing `}` you don't see right away that there is some `else-if` statement. But it's mostly one line of comment so it's not far away. I'm using such commenting style for already quite a long time and I've never had any problems with it. However, it would be worth checking how many people are using this style of commenting.

**Comment by waitingsong** [comment]
I prefer not to move the position of comments. It would show different meaning and include differenct scope at different position,  some for next line and some for a block (if/else...).

**Comment by dasa** [comment]
It seems like this Issue would be fixed by adding support for "break before else" (#840)

**Comment by lukejagodzinski** [comment]
@dasa I think the `breakBeforeElse` option is not needed here. As someone mentioned in the discussion there, prettier just shouldn't move comments when formatting. So, if someone puts a comment before else it wouldn't move it after else but leave where it was. Adding too many options will create yet another configuration hell beautifier.

**Comment by dasa** [comment]
I think you're referring to https://github.com/prettier/prettier/issues/840#issuecomment-284168624. I guess whichever is easier for `prettier`.

_(Not having this option is preventing my team from using `prettier` though.)_

**Comment by hyperknot** [comment]
Yes, just came to say that this is a 99.9% perfect project, really, except for this one extremely annoying behaviour.

Please allow me to write bigger if-else blocks like this:

```
// First condition comment.
if (firstCondition()) {
  doSomething();
}

// Second condition comment.
else if (secondCondition()) {
  doSomethingElse();
}
```

**Comment by suchipi** [comment]
@azz how do you feel about *not* putting `else` on the same line as `}`, but *only* if there's a comment in-between (otherwise we keep existing behavior)?

This is similar to respecting comments in import specifiers, eg:

**Prettier 1.11.1**
[Playground link](https://prettier.io/playground/#N4Igxg9gdgLgprEAuEBLAtgBwgJxgAmHwDMIJ8BfEnCdfAHRACMBDHRgbnqm4Cs4AzjAB06CGADWACkasAXowCUXKBmx5C3fPgD0O-IIA2qWAFoAJqgEsmhuKahwAHjFPHH+Nbhg6oLdHDmWvhikoEAEnCGmHDsUFTENHSyLAogHCAANCAQmDCo0ALIoGw0AO4ACmwIRSgsAG4QqOZZICxCyMQshgJw2Uw4LGEwAMqYQyYA5sgwOACufSDm4p3dvdkmvXgVg5PoLKs9i7wCTgBCg8Mj-nAAMiZwh+s5czCYrwBMT4vjOFvIzBsAE9DNBWpgcCYYAB1ZowAAWyAAHAAGbIQiC9aGDTAAiGCWL1R7ZQZQSZwACiUBaSDRIFJ5JGMDYMGQdJwcAAjnNUBydiw9gckF0jtleuhUDN5osBFM7ABFOYQeDfbLMpiw8wI5AfNWDVDuSYAYVo+wBUGgxJAc16ABUbLURb0KBQgA)

**Input:**
```jsx
import { foo } from "bar";

jest.mock("baz");
import {
  // eslint-disable-next-line import/named
  mockedHelper
} from "baz";
```

**Output:**
```jsx
import { foo } from "bar";

jest.mock("baz");
import {
  // eslint-disable-next-line import/named
  mockedHelper
} from "baz";
```

**Comment by duailibe** [comment]
@suchipi The general problem here is that we always attach a comment to a node and we currently can't attach as "leading comment of the else block"

**Comment by hyperknot** [comment]
I don't know the technical details, but I strongly agree with the "allow else on a separate line, IF there is a comment before it" idea. This way the classic behaviour doesn't change, yet everyone can separate important else switches to their own visual block.

**Comment by suchipi** [comment]
@duailibe can we attach it as the trailing comment of the consequent BlockStatement?

**Comment by duailibe** [comment]
@suchipi yep, that works! But it doesn't solve the case when the consequent is not a `BlockStatement`.

**Comment by suchipi** [comment]
I think a trailing comment works alright already in that case anyway:

**Prettier 1.11.1**
[Playground link](https://prettier.io/playground/#N4Igxg9gdgLgprEAuEBLAZgAgBSSgEwEpMAnOGAVxKgG5MB6ezaOAHSk0zgBsBnOUuSq0GTGAHcIIADQgIABxipovZKACGJEhHEAFTQlUp1ANwip8MkOt4xk6dXziyARiXVgA1uQDK8j6hQAObIMCQUziD4EGD2jvyygfwkMLruQQC26nFOsgBWvAAeAELuXr7qGXAAMoFwOQlyFDDyzQBMDZH+JMnIIC7qLgCe3NBW8iSBMADqFjAAFsgAHAAMshMQ-NPu8n0TcMkm9bLuwXAAogTIayCnQXA+MJp2SDdkAI4UqGRp6pnZSAcuRA-AyqFC4UivECQW4cAAihQIPBOrIni5ZvgFsg2mj3KhuDCAMIQDJZPpQFhWCj8AAqgyMQP4AF9mUA)
```sh
--range-end 
--range-start 
```

**Input:**
```jsx
if (cond) return; // one
  else return; // two
```

**Output:**
```jsx
if (cond)
  return; // one
else return; // two
```

**Comment by duailibe** [comment]
I believe people expect the following instad:

```js
if (cond) return;
// one
else
  return; // two
```

**Comment by suchipi** [comment]
Agreed

**Comment by hyperknot** [comment]
I didn't realise this is actually implemented in 1.12, which works wonderfully! Thank you so much, this really makes prettier so much pleasant experience in my opinion!
